Better npm post-install for virtual machines

xavierleune opened the pull request:
This pull requests resolves a bug in npm postinstall.

## Environment
- Virtual machine: Debian 8
- Hypervisor: Virtualbox 5.1.14
- Host OS: Windows 10
- Shared folders: SMB via vagrant

## Steps to reproduce
When you have a running environment, go to the shared folders where you cloned amphtml-validator. Then run `npm run postinstall`.

## Actual output

```
[08:18:43]-[/var/www/test-amp] $ npm i amphtml-validator

> amphtml-validator@1.0.19 postinstall /var/www/test-amp/node_modules/amphtml-va                                           lidator
> /bin/sh -c "exit 0" 2> NUL && rm NUL || node postinstall-windows.js

sh: 1: cannot create NUL: Permission denied
postinstall-windows.js: This script is for Windows only.
npm WARN enoent ENOENT: no such file or directory, open '/var/www/test-amp/packa                                           ge.json'
npm WARN test-amp No description
npm WARN test-amp No repository field.
npm WARN test-amp No README data
npm WARN test-amp No license field.
npm ERR! Linux 3.16.0-4-amd64
npm ERR! argv "/usr/local/node-v6.9.5-linux-x64/bin/node" "/usr/local/bin/npm" "                                           i" "amphtml-validator"
npm ERR! node v6.9.5
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE

npm ERR! amphtml-validator@1.0.19 postinstall: `/bin/sh -c "exit 0" 2> NUL && rm                                            NUL || node postinstall-windows.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the amphtml-validator@1.0.19 postinstall script '/bin/sh -c "                                           exit 0" 2> NUL && rm NUL || node postinstall-windows.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the amphtml-validator pac                                           kage,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     /bin/sh -c "exit 0" 2> NUL && rm NUL || node postinstall-windows.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs amphtml-validator
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls amphtml-validator
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /var/www/test-amp/npm-debug.log
```

## Expected output
Installation does not fails.

```
[08:40:28]-[/var/www/test-amp/node_modules/amphtml-validator] $ npm i

> amphtml-validator@1.0.19 postinstall /var/www/test-amp/node_modules/amphtml-validator
> node postinstall-windows.js

postinstall-windows.js: This script is for Windows only.
amphtml-validator@1.0.19 /var/www/test-amp/node_modules/amphtml-validator
âââ colors@1.1.2
âââ¬ commander@2.9.0
â âââ graceful-readlink@1.0.1
âââ¬ jasmine@2.3.2
â âââ exit@0.1.2
â âââ¬ glob@3.2.11
â â âââ inherits@2.0.3
â â âââ¬ minimatch@0.3.0
â â   âââ lru-cache@2.7.3
â â   âââ sigmund@1.0.1
â âââ jasmine-core@2.3.4
âââ¬ promise@7.1.1
  âââ asap@2.0.5
```

In this pull request we will run the `postinstall-windows.js` for every os and will exit without error when the current os is not Windows.

xavierleune commented:
I signed it!

xavierleune commented:
I signed it!

xavierleune commented:
It seems that travis build is failing for some reason, not depending of my changes...

jridgewell commented:
You'll have to rebase to get Travis to pass. Sorry about that.

jridgewell commented:
/to @powdercloud

xavierleune commented:
@jridgewell done ! Thank for your feedback

xavierleune commented:
@powdercloud tests are now running well ! What do you think about this pull request ?

jridgewell commented:
Is @powdercloud still out? /to @Gregable

powdercloud commented on validator/nodejs/package.json:
```diff
@@ -24,6 +24,6 @@
   },
   "scripts": {
-    "postinstall": "/bin/sh -c \"exit 0\" 2> NUL && rm NUL || node postinstall-windows.js"
+    "postinstall": "node postinstall-windows.js"
```
I think on Debian based systems (includes Ubuntu and LinuxMint) this will fail because the Node.js binary is called nodejs, not node. That's why this /bin/sh trickery exists...

xavierleune replied:
That was true 3 years ago. Now it's just node

powdercloud commented on validator/nodejs/postinstall-windows.js:
```diff
@@ -23,7 +23,7 @@ var fs = require('fs');
   console./*OK*/ error(
       'postinstall-windows.js: This script is for Windows only.');
-  process.exit(1);
+  process.exit(0);
```
I don't understand why we'd want to exit with success after deciding that we have an error just one line above.

xavierleune replied:
Because npm will apply this script on every platform and we decide to exit when non applicable. As long as we decide to launch the script, that's not an error anymore.

powdercloud commented:
Looking at the output you posted in your comment:

sh: 1: cannot create NUL: Permission denied

I wonder why this fails. Later in the output, we see that node_modules get created in your test directory, so it's not readonly. I wonder whether the creation of NUL gets prevented at the filesystem level (either by Windows or by the SMB share). I don't have a good idea for a fix yet ... just trying to understand it thus far.
I think the fix you posted won't work for some platforms unfortunately, I added a couple comments. The postinstall command line is a tricky thingy and we'll need to test whatever the solution is.

xavierleune commented:
Thanks for your feedback @powdercloud

Nul is invalid on Windows, Because that's like /dev/null on Linux. But on a debian, when I'm writing on a SMB share I'm actually on a Windows filesystem.
I dont see the whole point trying to do the detection on the command line, if node can do better.

xavierleune commented:
Output from a row ubuntu:
```
xleune@xleune-Latitude-E7470:~/Workspace/amphtml/validator/nodejs$ git checkout hotfix/node-postinstall
La branche hotfix/node-postinstall est paramétrée pour suivre la branche distante hotfix/node-postinstall depuis origin.
Basculement sur la nouvelle branche 'hotfix/node-postinstall'
xleune@xleune-Latitude-E7470:~/Workspace/amphtml/validator/nodejs$ npm install
npm WARN deprecated minimatch@0.3.0: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue

> amphtml-validator@1.0.19 postinstall /home/xleune/Workspace/amphtml/validator/nodejs
> node postinstall-windows.js

postinstall-windows.js: This script is for Windows only.
amphtml-validator@1.0.19 /home/xleune/Workspace/amphtml/validator/nodejs
├── colors@1.1.2 
├─┬ commander@2.9.0 
│ └── graceful-readlink@1.0.1 
├─┬ jasmine@2.3.2 
│ ├── exit@0.1.2 
│ ├─┬ glob@3.2.11 
│ │ ├── inherits@2.0.3 
│ │ └─┬ minimatch@0.3.0 
│ │   ├── lru-cache@2.7.3 
│ │   └── sigmund@1.0.1 
│ └── jasmine-core@2.3.4 
└─┬ promise@7.1.1 
  └── asap@2.0.5 
```
Seems ok, I'll run tests on mac os and windows too

xavierleune commented:
@powdercloud after thinking this with a fresh eye, I'm proposing a new approach. The key fact is the problem come from the file "NUL". So here is the actual process:
Command line: `/bin/sh -c \"exit 0\" 2> NUL && rm NUL || node postinstall-windows.js`

* **Raw Windows**: the first part `/bin/sh` will fail and the error will be redirected to NUL. Then we try to `rm` the file `NUL` (which doesn't work in cmd) and the `||` will launch the postinstall
* **Cygwin**: Despite the fact that nodejs [does not support cygwin anymore](https://github.com/nodejs/node/wiki/Installation#building-on-cygwin). The fact here is not to launch the postinstall when on cygwin (according to comments in the main script)

I've tested my new approach under a raw ubuntu, debian over CIFS, raw macOs and cmd under Windows 10. It works in any case. I've not tested it under cygwin, because the latest version I can compile is way too old (0.4.10).

Does that new approach satisfies you more ?

powdercloud commented on validator/nodejs/package.json:
```diff
@@ -24,6 +24,6 @@
   },
   "scripts": {
-    "postinstall": "/bin/sh -c \"exit 0\" 2> NUL && rm NUL || node postinstall-windows.js"
+    "postinstall": "/bin/sh -c \"exit 0\" 2> output && rm output || (del output && node postinstall-windows.js)"
```
Rather than calling it 'output', may be better to make it more unique, like postinstall.DELETEME, to avoid that it clobbers something useful in the current working directory.
As for the del command, perhaps it's better to delete the file inside postinstall-windows.js? The del command will fail on non-Windows systems, so we could never get to the error message inside the script.

I think overall this solution is getting there, thanks a lot for looking into it.

xavierleune commented:
@powdercloud thanks ! I changed the name of the temp file and moved deletion for windows to the postinstall script. The postinstall does not fails if the file does not exists but a message is logged if we could not delete the file.

powdercloud commented on validator/nodejs/postinstall-windows.js:
```diff
@@ -20,6 +20,18 @@
+	}
+});
+
+fs.exists('', function(exists) {
```
The '' here looks like a leftover / unintentional? Maybe just do the sync version as a best effort here for simplicity. I think it's fine to skip the error logging, in the very rare and unexpected case something goes wrong the function will throw an exception. E.g. this (I haven't tested this yet but we have a Windows box sitting around in another building for gaming ... I promise I will wander over there and see how this works ...):

// On Windows, the script invocation (see package.json's scripts.postinstall line)
// will leave this temporary file behind, so we make an effort to clean it up.
if (fs.existsSync('postinstall.DELETEME')) {
  fs.unlinkSync('postinstall.DELETEME');
}

xavierleune replied:
I think I should stop working on a saturday :smile: I'll do the changes you requested
